lib/services/namespace_service.ts

```diff
@@ -0,0 +1,139 @@
+import { PureService } from '@Services/pure_service';
+import { DeviceInterface } from '@Lib/device_interface';
+import { Uuid } from '@Lib/uuid';
+import { UuidString } from './../types';
+
+export type SNNamespace = {
+  identifier: string | UuidString,
+  userUuid?: UuidString,
+  label: string,
+  isDefault: boolean,
+};
+
+const NAMESPACES_STORAGE_KEY = 'namespaces';
+
+/**
+ * The namespace service is responsible of setting the namespace to be used by
+ * the DeviceInterface.
+ */
+export class SNNamespaceService extends PureService {
+  private namespace?: SNNamespace
+
+  constructor(deviceInterface: DeviceInterface) {
+    super();
+    this.deviceInterface = deviceInterface;
+  }
+
+  private async getNamespaces() {
+    const namespaces = await this.deviceInterface!.getRawStorageValue(NAMESPACES_STORAGE_KEY);
+    if (!namespaces) {
+      return [] as SNNamespace[];
+    }
+    return JSON.parse(namespaces) as SNNamespace[];
+  }
+
+  private async setNamespaces(namespaces: SNNamespace[]) {
+    this.deviceInterface!.setRawStorageValue(
+      NAMESPACES_STORAGE_KEY,
+      JSON.stringify(namespaces)
+    );
+  }
+
+  private async getDefaultNamespace() {
+    const namespaces = await this.getNamespaces();
+    return namespaces.find(namespace => namespace.isDefault === true);
+  }
+
+  private async pushNamespace(namespace: SNNamespace) {
+    const namespaces = await this.getNamespaces();
+    const namespaceIndex = namespaces.findIndex(n => n.identifier === namespace.identifier);
+    if (namespaceIndex === -1) {
+      namespaces.push(namespace);
+    } else {
+      namespaces[namespaceIndex] = namespace;
+    }
+    await this.setNamespaces(namespaces);
+  }
+
+  private async setCurrentNamespace(namespace: SNNamespace) {
+    this.namespace = namespace;
+    this.deviceInterface!.switchToNamespace(namespace);
+    await this.pushNamespace(namespace);
+  }
+
+  private async createNamespace(isDefault: boolean = false, label?: string) {
+    if (isDefault) {
```

Comment:
Can be simplified to `if (isDefault && await this.getDefaultNamespace()) { ...exception...}`